ingestion/src/metadata/workflow/base.py

```diff
@@ -0,0 +1,245 @@
+from metadata.ingestion.api.step import Step
+from metadata.ingestion.api.steps import BulkSink, Processor, Sink, Source, Stage
+from metadata.ingestion.models.custom_types import ServiceWithConnectionType
+from metadata.ingestion.ometa.ometa_api import OpenMetadata
+from metadata.timer.repeated_timer import RepeatedTimer
+from metadata.utils.class_helper import (
+    get_service_class_from_service_type,
+    get_service_type_from_source_type,
+)
+from metadata.utils.logger import ingestion_logger, set_loggers_level
+from metadata.workflow.workflow_output_handler import get_ingestion_status_timer
+from metadata.workflow.workflow_status_mixin import WorkflowStatusMixin
+
+logger = ingestion_logger()
+
+T = TypeVar("T")
+
+REPORTS_INTERVAL_SECONDS = 60
+
+
+class InvalidWorkflowJSONException(Exception):
+    """
+    Raised when we cannot properly parse the workflow
+    """
+
+
+class BaseWorkflow(ABC, WorkflowStatusMixin):
+    """
+    Base workflow implementation
+    """
+
+    config: OpenMetadataWorkflowConfig
+    _run_id: Optional[str] = None
+    service_type: ServiceType
+    metadata_config: OpenMetadataConnection
+    metadata: OpenMetadata
+
+    # All workflows require a source as a first step
+    source: Source
+    # All workflows execute a series of steps, aside from the source
+    steps: Tuple[Step]
+
+    def __init__(self, config: OpenMetadataWorkflowConfig):
+        """
+        Disabling pylint to wait for workflow reimplementation as a topology
+        """
+        self.config = config
+        self._timer: Optional[RepeatedTimer] = None
+
+        set_loggers_level(config.workflowConfig.loggerLevel.value)
+
+        self.service_type: ServiceType = get_service_type_from_source_type(
+            self.config.source.type
+        )
+
+        self.metadata_config: OpenMetadataConnection = (
+            self.config.workflowConfig.openMetadataServerConfig
+        )
+        self.metadata = OpenMetadata(config=self.metadata_config)
+
+        self.set_ingestion_pipeline_status(state=PipelineState.running)
+
+        # Pick up the service connection from the API if neededweb_analytic_report_data_processor
+        self._retrieve_service_connection_if_needed(self.service_type)
+
+        # Informs the `source` and the rest of `steps` to execute
+        self.set_steps()
+
+    @abstractmethod
+    def set_steps(self):
+        """
+        initialize the tuple of steps to run for each workflow
+        and the source
+        """
+
+    def _execute_internal(self):
+        """
+        Internal execution that needs to be filled
+        by each ingestion workflow.
+
+        Pass each record from the source down the pipeline:
+        Source -> (Processor) -> Sink
+        or Source -> (Processor) -> Stage -> BulkSink
+
+        Note how the Source class needs to be an Iterator. Specifically,
+        we are defining Sources as Generators.
+        """
+        for record in self.source.run():
```

Comment:
where do we enforce that `Source` should have a run()

Reply:
[here](https://github.com/open-metadata/OpenMetadata/pull/13029/files#diff-0299422e9540c993001cd05eea601d12500c7f3de9b589d596625458e0346260R144). Might add the `run` as an abstractmethod to the Step class itself

Reply:
the `Source` class inherit from the `StageStep` class?